packages/block-library/src/table-of-contents/utils.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * WordPress dependencies
+ */
+import { select } from '@wordpress/data';
+import { cleanForSlug } from '@wordpress/editor';
+
+export function linearToNestedList( array, index = 0 ) {
+	const returnValue = [];
+
+	array.forEach( function( heading, key ) {
+		if ( heading.content === undefined ) {
+			return;
+		}
+
+		// Make sure we are only working with the same level as the first iteration in our set.
+		if ( heading.level === array[ 0 ].level ) {
+			// Check that the next iteration will return a value.
+			// If it does and the next level is greater than the current level,
+			// the next iteration becomes a child of the current interation.
+			if (
+				array[ key + 1 ] !== undefined &&
+				array[ key + 1 ].level > heading.level
+			) {
+				// We need to calculate the last index before the next iteration that has the same level (siblings).
+				// We then use this last index to slice the array for use in recursion.
+				// This prevents duplicate nodes.
+				let endOfSlice = array.length;
+				for ( let i = key + 1; i < array.length; i++ ) {
+					if ( array[ i ].level === heading.level ) {
+						endOfSlice = i;
+						break;
+					}
+				}
+
+				// We found a child node: Push a new node onto the return array with children.
+				returnValue.push( {
+					block: heading,
+					index: index + key,
+					children: linearToNestedList(
+						array.slice( key + 1, endOfSlice ),
+						index + key + 1
+					),
+				} );
+			} else {
+				// No child node: Push a new node onto the return array.
+				returnValue.push( {
+					block: heading,
+					index: index + key,
+					children: null,
+				} );
+			}
+		}
+	} );
+
+	return returnValue;
+}
+
+export function getPageHeadings() {
+	return convertHeadingBlocksToAttributes( getHeadingBlocks() );
+}
+
+export function getHeadingBlocks() {
+	const editor = select( 'core/block-editor' );
+	return editor
+		.getBlocks()
+		.filter( ( block ) => block.name === 'core/heading' );
+}
+
+export function convertHeadingBlocksToAttributes( headingBlocks ) {
+	return headingBlocks.map( function( heading ) {
+		const level = heading.attributes.level.toString();
```

Comment:
Why is `level` a string instead of a number?

Reply:
This would probably be because it's sourced from an HTML attribute. Block attributes that are sourced from HTML attributes have to be strings.

Reply:
@talldan [But it's not sourced from an HTML attribute, and it's stored as a number.](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/heading/block.json#L15)

Reply:
@ZebulanStanphill The value comes from the heading block as a `number`, but is converted to a `string` here so that it can be stored/sourced as an attribute in the table of contents block using a data attribute:
https://github.com/WordPress/gutenberg/pull/21040/files#diff-ca24950398fffc6e682a847d2d23f966R31

Reply:
Aha! nice, but at that position in code it's hard to figure. Maybe convert later before assignment ...

Reply:
Yep, a comment explaining the conversion might also be a solution.